dnprogs/apps/ctermd.8
.TH CTERMD 8 "July 27 1998" "DECnet utilities"

.SH NAME
ctermd  \- CTERM services for Linux  

.SH SYNOPSIS
.B ctermd 
.br
[options]
.br
Options:
.br
[\-dvVh] [\-l logtype]
.SH DESCRIPTION
.PP
Allows remote users to connect as a terminal over DECnet.
.br
 
.br
This application implements the CTERM protocol over DECnet for allowing a 
remote DECnet host emulating a DEC VT100 terminal to connect to Linux.
.br

.br
Normally this daemon will be run from dnetd(8) rather than by hand.

.SH OPTIONS
.TP
.I "\-l"
Set logging options. The following are available:
.br
.B -lm
Log to /dev/mono. (only useful if you have my mono monitor driver or mdacon
and a second monitor)
.br
.B -le
Log to stderr. Use this for debugging or testing combined with
.B -d.
.br
.B -ls
Log to syslog(3). This is the default if no options are given.
.TP
.I "\-d"
Don't fork and run the background. Use this for debugging.
.TP
.I "\-v"
Verbose. The more of these there are the more verbose ctermd will be.
.TP
.I \-h \-?
Displays help for using the command.
.TP
.I \-V
Show the version of ctermd.

.SH EXAMPLES
.br
Starting the ctermd daemon/Connecting to linux from a VAX.
.br

.br
.PP
On Linux:

   # ctermd  

On the VAX:

   $ set host pclnx

   Ctermd Version 1.0.0
   DECnet for Linux

   login:

.br

.SH SEE ALSO
.BR dntype "(1), " dndir "(1), " dndel "(1), " dntask "(1), "
.BR dnetd "(8), " dnping "(1), " sethost "(1), " dnetd.conf (5)


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/apps/ctermd.c
/******************************************************************************
    (c) 1995-2002 E.M. Serrat             emserrat@geocities.com

    Conversion to dnetd (c) 1999 by Christine Caulfield

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*******************************************************************************/

#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <signal.h>
#include <string.h>
#include <fcntl.h>
#include <syslog.h>
#include <ctype.h>
#include <regex.h>
#include <stdlib.h>
#include <utmp.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>
#ifdef DNETUSE_DEVPTS
#include <pty.h>
#endif

#include "dn_endian.h"

static	struct	sockaddr_dn		sockaddr;
static	char				*line;
static	int				s,t,net,pty,len;
static	int				read_present;
/*-----------------------------------------------------------------------*/
void cterm_child(int s)
{
	int	sts;
	wait (&sts);
}
/*-----------------------------------------------------------------------*/
void cterm_setchar(void)
{
	char	cterm_setchar_msg[] = {	0x09,0x00,	/* common data 	*/
					0x0f,0x00,	/* Length	*/
					0x0b,0x00,
					0x06,0x02,	/* Input ESC seq*/
							/* Recognition	*/
					0x01,0x00,
					0x07,0x02,	/*output ESC seq*/
							/* Recognition	*/
					0x01,0x00,

					0x02,0x02,0x19,
					0xff,0x00
				      };

	if (write(net,cterm_setchar_msg,sizeof(cterm_setchar_msg)) <  0)
	{
		perror("Cterm setchar");
		exit(-1);
	}
}
/*-----------------------------------------------------------------------*/
void cterm_bind(void)
{
	char	cterm_bind_msg[] = {0x01,0x02,0x04,0x00,0x07,0x00,0x10,0x00};
	char	cterm_init_msg[] = {0x09,0x00, 		/* common data */
				    0x1b,0x00,		/* Block length*/
				    0x01,0x00,		/* init messg */
				    0x01,0x04,0x00,	/* Version     */
				    0,0,0,0,0,0,0,0,    /* descrip     */
				    0x01,0x02,0x00,0x02,
				    0x02,0x02,0xf4,0x03,
				    0x03,0x04,0xfe,0x7f,0x00,0x00};
	char	buf[512];

	if (write(net,cterm_bind_msg,sizeof(cterm_bind_msg)) < 0)
	{
		perror("Cterm bind request failed");
		exit (-1);
	}

	if (read(net,buf,sizeof(buf)) < 0)
	{
		perror("Cterm bind accept failed");
		exit (-1);
	}

	if (buf[0] != 0x04)
	{
		printf("Error Setting up cterm link\n");
		exit(-1);
	}

	if (write(net,cterm_init_msg,sizeof(cterm_init_msg)) < 0)
	{
		perror("Cterm init message failed");
		exit (-1);
	}
	if (read(net,buf,sizeof(buf)) < 0)
	{
		perror("Cterm init receive failed");
		exit (-1);
	}
	cterm_setchar();
}
/*-----------------------------------------------------------------------*/
void cterm_unbind (void)
{
	char	cterm_unbind_msg[3] = {0x02,0x03,0x00};
	write(net,cterm_unbind_msg,3);
}
/*-----------------------------------------------------------------------*/
void cterm_reset(int s)
{
	struct	utmp	entry;
	struct	utmp	*lentry;
	char	*p;

	cterm_unbind();
	sleep(1); /* ugh!. Wait for data-ack to arrive */
#ifndef DNETUSE_DEVPTS
	p=line+sizeof("/dev/")-1;

	setutent();
	memcpy(entry.ut_line,p,strlen(p));
	entry.ut_line[strlen(p)]='\0';
	lentry=getutline(&entry);
	lentry->ut_type=DEAD_PROCESS;

	memset(lentry->ut_line,0,UT_LINESIZE);
	memset(lentry->ut_user,0,UT_NAMESIZE);
	lentry->ut_time=0;
	pututline(lentry);

	(void)chmod(line,0666);
	(void)chown(line,0,0);
	*p='p';
	(void)chmod(line,0666);
	(void)chown(line,0,0);
#endif
	close(pty);
	close(net);
	exit(1);
}
/*-----------------------------------------------------------------------*/
void cterm_purge(void)
{
	char	buf[4000];
	char	Control_c = 0x03;
	long	numbytes;

	write(pty,&Control_c,1);
	ioctl(pty,FIONREAD,&numbytes);
	while (numbytes)
	{
	 	read(pty,buf,numbytes);
		ioctl(pty,FIONREAD,&numbytes);
	}
}
/*-----------------------------------------------------------------------*/
void cterm_unread (void)
{
	char	cterm_unread_msg[] = {	0x09,0x00,0x02,0x00,0x05,0x00 };
	write(net,cterm_unread_msg,sizeof(cterm_unread_msg));
	read_present=0;
}
/*-----------------------------------------------------------------------*/
void cterm_read (void)
{
	char	cterm_stread_msg[] = {	0x09,0x00,0x31,0x00,
					0x02,
					0x40,0x4A,0x02,
					0xa0,0x00,0x00,0x00,
					0x00,0x00,0x00,0x00,
					0x00,0x00,0x00,0x00,
					0x20,
					0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
					0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
					0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
					0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
				     };
	write(net,cterm_stread_msg,sizeof(cterm_stread_msg));
	read_present=1;
}
/*-----------------------------------------------------------------------*/
void cterm_write (char *buf)
{
	char	cterm_write_msg[] = {	0x09,0x00,
					0x00,0x00,
					0x07,0x30,0x00,0x00,0x00};
	char	lclbuf[1400];
	short	*blklen;
	int	wrtlen;
	int     l;

	memcpy(&lclbuf[0],cterm_write_msg,9);
	blklen=(short *)&lclbuf[2];
	l = 5 + strlen(buf);
	lclbuf[2] = l & 0xFF;
	lclbuf[3] = (l>>8) & 0xFF;

	memcpy(&lclbuf[9],buf,strlen(buf));
	wrtlen=strlen(buf)+9;

	if (write(net,lclbuf,wrtlen) < 0)
	{
		perror("cterm_write");
		exit(-1);
	}

}
/*-----------------------------------------------------------------------*/
void cterm (void)
{
	char	 buf[100];
	fd_set	 rdfs;
	int	 cnt;
	struct   sigaction siga;
	sigset_t ss;

	setsid();

	sigemptyset(&ss);
	siga.sa_handler=cterm_reset;
	siga.sa_mask  = ss;
	siga.sa_flags = 0;
	sigaction(SIGCHLD, &siga, NULL);

	signal(SIGTSTP, SIG_IGN);
	signal(SIGTTOU, SIG_IGN);

	cterm_write("CTERM Version 1.0.6\r\nDECnet for Linux\r\n\n");
	cterm_read();

	for (;;)
	{
		FD_ZERO(&rdfs);
		FD_SET(pty,&rdfs);
		FD_SET(net,&rdfs);

		 if (select(FD_SETSIZE,&rdfs,NULL,NULL, NULL) > 0)
		 {
		  if (FD_ISSET(pty,&rdfs))
		  {
			cnt=read(pty,buf,sizeof(buf)-1);
			if (cnt <= 0) break;
			buf[cnt]='\0';
			cterm_write(buf);
		   }

		   if (FD_ISSET(net,&rdfs))
		   {
			cnt=read(net,buf,sizeof(buf));
			if (cnt <= 0) break;

			switch (buf[4])
			{
		 	case 0x03:
				if ( (buf[12] == 0x19) || (buf[12] == 3) )
				{
					buf[12]=0x03;
					cterm_purge();
				}
				write(pty,&buf[12],cnt-12);
				cterm_read();
				break;
		 	case 0x04:
				if ( (buf[6] == 0x19) || (buf[6] == 3) )
				{
					buf[6]=0x03;
					cterm_purge();
				}
				write(pty,&buf[6],1);
				cterm_read();
			}
		     }
		    }
	}
	cterm_reset(0);
}
/*-----------------------------------------------------------------------*/
void doit  (void)
{
	int	i,c;
	struct	stat	stb;
	char	ptyname[] = "/dev/ptyCP";
	int	gotpty =0;

	cterm_bind();

#ifdef DNETUSE_DEVPTS
 	if (openpty(&pty, &t,NULL, NULL, NULL) == 0)
	  gotpty = 1;
	else
	{
	    cterm_write("No ptys available for connection");
	    exit(-1);
	}
#else
	for (c='p'; c <= 'z'; c++)
	{
		line = ptyname;
		line[strlen("/dev/pty")] = c;
		line[strlen("/dev/ptyC")] = '0';
		if (stat(line,&stb) < 0)
			break;
		for (i=0; i < 16; i++)
		{
			line[strlen("/dev/ptyC")]= "0123456789abcdef"[i];
			if ( (pty=open(line,O_RDWR)) > 0)
			{
				gotpty = 1;
				break;
			}
		}
		if (gotpty) break;
	}
	if (!gotpty)
	{
		cterm_write("No ptys available for connection");
		exit(-1);
	}


	line[strlen("/dev/")] = 't';
	if ( (t=open(line,O_RDWR)) < 0)
	{
		cterm_write("Error connecting to physical terminal");
		exit(-1);
	}
#endif
	if ( fchmod(t,0) )
	{
		cterm_write("Error setting terminal mode");
		exit(-1);
	}

	if ( ( i=fork() ) < 0)
	{
		cterm_write("Error forking");
		exit(-1);
	}
	if (i)
	{
		cterm();
	}

	setsid();

	close(pty); close(net);
	if (t != 0) dup2 (t,0);
	if (t != 1) dup2 (t,1);
	if (t != 2) dup2 (t,2);

	if (t > 2) close(t);

	ioctl(0, TIOCSCTTY, (char *)NULL);


	putenv("TERM=vt100");
	execlp("/bin/login", "login", (char *)0);
	cterm_write("Error executing login command");
	exit(-1);
}

void usage(char *prog, FILE *f)
{
    fprintf(f,"\n%s options:\n", prog);
    fprintf(f," -v        Verbose messages\n");
    fprintf(f," -h        Show this help text\n");
    fprintf(f," -d        Run in debug mode - don't do initial fork\n");
    fprintf(f," -l<type>  Logging type(s:syslog, e:stderr, m:mono)\n");
    fprintf(f," -V        Show version number\n\n");
}


/*-----------------------------------------------------------------------*/
int main(int argc, char *argv[])
{
    int verbosity;
    int debug = 0;
    char log_char = 's';
    char opt;

    /* Deal with command-line arguments.*/
    opterr = 0;
    optind = 0;
    while ((opt=getopt(argc,argv,"?vVdhl:")) != EOF)
    {
	switch(opt)
	{
	case 'h':
	    usage(argv[0], stdout);
	    exit(0);

	case '?':
	    usage(argv[0], stderr);
	    exit(0);

	case 'v':
	    verbosity++;
	    break;

	case 'd':
	    debug++;
	    break;

	case 'V':
	    printf("\nctermd from dnprogs version %s\n\n", VERSION);
	    exit(1);
	    break;

	case 'l':
	    if (optarg[0] != 's' &&
		optarg[0] != 'm' &&
		optarg[0] != 'e')
	    {
		usage(argv[0], stderr);
		exit(2);
	    }
	    log_char = optarg[0];
	    break;

	}
    }

    /* Initialise logging */
    init_daemon_logging("ctermd", log_char);

    net = dnet_daemon(DNOBJECT_CTERM, NULL, verbosity, !debug);

    if (net > -1)
    {
	dnet_accept(net, 0, NULL, 0);
	doit();
    }
    return 0;
}
/*-------------------------------------------------------------------------*/

